### ISSUE TITLE
Shipping is a feature

### ISSUE THREAD

**Issue by paulbrucecotton** [w3c/wg-effectiveness#12]
As Chair of the HME WG I had to convince the Editors and the most active HME WG participants that "shipping is a feature".  ALL W3C WG member need to understand this concept.

By this I mean "shipping a specification" (eg. getting it to CR or PR by a particular date) should be treated like any other feature in the specification and the "shipping feature" should be treated on an equal footing with any other feature requests.  All too often WG members want to add new features to a specification at the cost of the "shipping feature" and they do not understand the impact of doing this.

As a WG progresses along its chartered timeline the "shipping feature" starts to become more and more important.

For example, in early 2016 I had to inform the HME WG participants that we ran the risk of running out of charter time to finish our MSE and EME specifications.  I basically had to convince the WG participants that the only feature that mattered at that point in time was the "shipping feature" or else all the other features would be lost if the WG was closed without finishing its deliverables.  The Editors then started a [triage process](https://lists.w3.org/Archives/Public/public-html-media/2016Mar/0003.html) to move the specs along.

Eventually we did ship both MSE and EME but I was surprised how hard it was to convince HME WG participants that "shipping is a feature" even though this culture is very prevalent in many companies that ship software products.

**Comment by nrooney** [comment]
Thanks for this - the "Triage Process" is something which we may want to explore, and part of that could be working through the CSI document to find out why "Shipping is a Feature" has not been adhered to. Other then that I wonder whether there is a culture for striving for perfection or for a scope rich spec which people may believe is more important than shipping on time. I wonder how others feel about this...

All - how do you feel about the conflicting situation when group members want to expand specs which will delay the spec publication date? Is this something which we should give advice on, or should be managed in a spec-by-spec basis?

**Comment by marcoscaceres** [comment]
We've been successfully following the "shipping is a feature" model with Payment Request, but the problem is largely that people don't fully understand what shipping a specification actually means. 

To clarify: if  a group ships a specification without a test suite, they basically have just a bunch of words and very little chance of interop (and more than likely security and spec bugs). So shipping in such a case is going to cause more harm than good. 

So, "shipping is a feature" actually constitutes a couple of things, which @paulbrucecotton touched on to a degree:

 * Scope: reduced feature set to things that are critical for new web platform feature to address the most critical use cases (initially leave out the nice to haves, but incubate  those in parallel if need be).
 * Tests: reduced feature set must be tested. 
 * Commitment to actually write the tests: this one is hard, because a lot of people who work on specs for JavaScript APIs don't actually know JavaScript/WebIDL (😱) or the web security model (🤢) or sometimes even HTML (😔).

For the last point, we need to do a better job at getting people skilled up to participate effectively - otherwise, WGs risk becoming talk shops, or things that don't make sense end up getting proposed (and rejected, which wastes valuable time, causes friction and frustration in the community, etc.). 

For payment request, we've been fairly strict about the reduced feature set, and have deferred a bunch of things to future spec work/features. See our [list of future features](https://github.com/w3c/payment-request/issues?q=is%3Aissue+is%3Aopen+label%3A%22Future+Candidate+Feature%22) to get a concrete idea of what I mean. Compare that to our list of things we need to do to [exit CR](https://github.com/w3c/payment-request/issues?q=is%3Aissue+is%3Aopen+label%3ACR-Exit). 

Having said that, feature development, and incubation of new features should follow the the [WHATWG Work Mode](https://whatwg.org/working-mode). That is, features should be incubated, debated, tested, implemented, and then brought into a spec proper.  

What is nice from the above is that it's possible to incrementally develop features, one at a time, fully tested, and incrementally ship them. Also, nothing unstable ever ends up in a spec, because features have been fully vetted before they ship.

**Comment by nrooney** [comment]
Oh I like the "list of future features" and "exit CR" tabs on issues in gh, keeps people focused on what needs to get done. Not sure where these fit in our outcomes at the moment so will discuss on the call today (I like how payment request uses these, but not sure all groups use gh, suppose in that case wgs could just setup a wiki page for future features and exit CR issues).

The WHATWG work mode topic should be discussed too; although perhaps as another topic as it probably will instigate a few more conversations!

**Comment by marcoscaceres** [comment]
> not sure all groups use gh

Perhaps part of making groups efficient could involve encouraging to use gh? It provides very good tools. 

It would be interesting to see which groups haven't migrated and why (and if they are doing effective work)?

**Comment by dontcallmedom** [comment]
there are hardly any groups that haven't switched to github. looking at the [report of my script validating repos](https://w3c.github.io/validate-repos/report.html), and excluding WGs that are dormant, the only groups that haven't switched to github are EXI, Tracking Protection and WebFonts.

**Comment by astearns** [comment]
I think testing probably deserves a separate issue as it would be good to tease out suggestions specific to that topic.
On shipping, I'd note that supergroups (like CSS) can't rely on re-chartering as a deadline to focus minds, as they almost always get re-chartered to continue the work that doesn't get to PR. One mechanism that does work for CSS is module levels. Once a module gets close to CR it becomes easier to shift newer and less-tested ideas to a future level and reduce scope on the current level.

**Comment by marcoscaceres** [comment]
> the only groups that haven't switched to github are EXI, Tracking Protection and WebFonts.

That's great to hear.

**Comment by marcoscaceres** [comment]
@astearns, I filed a new bug related to testing https://github.com/w3c/wg-effectiveness/issues/21

**Comment by nrooney** [comment]
Added deliverables related to this discussion

**Comment by jeffjaffe** [comment]
We should combine this with the Continuous Incubation (CI) document.

The whole point of CI is to not allow CI to stop us from shipping because there is always something new being added.  If we had a disciplined process to prevent CI from getting in the way - we would ship more often.  So let's adopt "shipping is a feature" as guidance to prevent CI stagnation.
